### PR TITLE
@callback can now be used directly (without Behaviour). Add @macrocallback.

### PR DESCRIPTION
**Pull request by antipax**
Implements (at least part of!) #3465.

* Updates existing `@callback` attr to attach `@doc` and raise on invalid use of defaults in typespec
* Adds new `@macrocallback` attr, which can be used to define behaviour macro callbacks. Also supports `@doc` and raise on invalid use of defaults in typespec
* Refactors `Behaviour` as a thin layer above the "new" `@callback` and `@macrocallback` attrs
* Adds `Kernel.Typespec.beam_callbackdocs/1`, which returns all callback docs from the given module's beam code.

### CODE REVIEW

File: lib/elixir/lib/kernel/typespec.ex [hunk 1]
```diff
@@ -475,14 +507,94 @@ defmodule Kernel.Typespec do
...
+    raise ArgumentError, "default arguments \\\\ not supported in @callback/@macrocallback"
+  end
+
+  defp ensure_not_default(_), do: :ok
+
   ## Macro callbacks
 
   @doc false
+  def defspec(:callback, expr, caller) do
+    {spec, _, _} = split_spec(expr, quote(do: term))
```
Comment (josevalim) [inline lib/elixir/lib/kernel/typespec.ex]: Since we don't do those checks for `@spec`, we should not do it for `@callback`. Or, alternatively, we should do it on all of them. My suggestion: keep this defspec as tiny as possible (don't split the spec nor check the arguments here). Store macrocallback still as macrocallback and process those just later on here:

https://github.com/antipax/elixir/blob/behaviour-enhancements-3465/lib/elixir/lib/kernel/typespec.ex#L662

In other words, inside that function you will check if `kind == :macrocallback` and convert the kind to callback, change the name and add the extra argument (`Macro.Env.t`) there.

**Comment by josevalim** [comment]
This is very good! We are almost there! The good news is that now we only need to make it simpler. :) I have added a comment.

**Comment by antipax** [comment]
@josevalim awesome, sounds good to me! Hope to wrap this up this evening or tomorrow.

File: lib/elixir/lib/kernel/typespec.ex [hunk 2]
```diff
@@ -550,6 +607,22 @@ defmodule Kernel.Typespec do
   defp translate_spec(kind, {:::, meta, [{name, _, args}, return]}, guard, caller) when is_atom(name) and name != ::: do
     if is_atom(args), do: args = []
 
+    if kind == :macrocallback do
+      kind = :callback
+      name = :"MACRO-#{name}"
+      args = [quote(do: env :: Macro.Env.t)|args]
+    end
+
+    :lists.foreach fn
```
Comment (josevalim) [inline lib/elixir/lib/kernel/typespec.ex]: Let's move this to a private function. "validate_no_defaults!" or something!

File: lib/elixir/test/elixir/kernel/typespec_test.exs [hunk 3]
```diff
@@ -642,4 +642,54 @@ defmodule Kernel.TypespecTest do
...
+    @callback literal(123, {atom}, :atom, [integer], true) :: atom
+
+    @macrocallback last(integer) :: Macro.t
+  end
+
+  test :callbacks do
+    assert Sample.behaviour_info(:callbacks) == [first: 1, guarded: 1, "MACRO-last": 2, literal: 5, orr: 1, foo: 2, bar: 2]
+  end
+
+  test :specs do
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/typespec_test.exs]: We can kill this test. :)

File: lib/elixir/test/elixir/kernel/typespec_test.exs [hunk 4]
```diff
@@ -642,4 +642,54 @@ defmodule Kernel.TypespecTest do
...
+    assert length(Keyword.get_values(Sample.module_info[:attributes], :callback)) == 7
+  end
+
+  test :default_is_not_supported do
+    assert_raise ArgumentError, fn ->
+      defmodule WithDefault do
+        @callback hello(num \\ 0 :: integer) :: integer
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/typespec_test.exs]: Let's use `@spec` in some of those assertions just to prove the point they are not allowed anywhere. :)

**Comment by josevalim** [comment]
:heart:
